### ISSUE TITLE
add extra tags to <head> per page

### ISSUE THREAD

**Issue by pontius-g** [vuejs/vuepress#894]
Hi

Help me please!
I there is a way to add custom tags into '\<head\>' section per page basis?
In example, adding <link rel='canonical' ...> ?

**Comment by ekoeryanto** [comment]
currently, there is no easy way to do that. 
but I'll check if i can create a plugin for that, maybe `vuepress-plugin-canonical`

**Comment by pontius-g** [comment]
Thanks, i made some "dirty fix" as per page script
```
<script>export default {created () { this.$ssrContext.userHeadTags+=`<link rel='canonical' href='${this.$page.path}'/>`;}}</script>
```

But I'll be appreciate for more "elegant" solution :)

**Comment by santigarcor** [comment]
@pontius-g I'm using v0.14 where did you place that script?

**Comment by pontius-g** [comment]
> @pontius-g I'm using v0.14 where did you place that script?

With a few correction, placed right to the bottom of .md file, where it’s needed

```
<script>export default {created () { if (typeof this.$ssrContext !== 'undefined') this.$ssrContext.userHeadTags+=`<link rel='canonical' href='https://rnd.tips${this.$page.path}'/>`;}}</script>
```

**Comment by ekoeryanto** [comment]
@pontius-g it's easier when the theme is ejected

**Comment by pontius-g** [comment]
I didn’t eject the theme to stay covered with further updates implicitly

**Comment by fatihtoprak** [comment]
it doesnt work for me ?

**Comment by alekspetrov** [comment]
Try this https://vuepress.vuejs.org/config/#head also you can add additional meta tags to the head of your site

**Comment by ulivz** [comment]
We're closing this issue as stale as it's more than 20 days without activity, and there are many workarounds as above.

**Comment by IOriens** [comment]
I made a [plugin](https://github.com/IOriens/vuepress-plugin-canonical) according previous discussions on this issue.

Just try
```shell
npm i vuepress-plugin-canonical -D
```
then modify `.vuepress/config.js`.

```javascript
module.exports = {
  plugins: [
    [
      'vuepress-plugin-canonical',
      {
        baseURL: 'https://mina.wiki', // base url for your canonical link, optional, default: ''
        stripExtension: true // strip '.html' , optional, default: false
      }
    ]
  ]
}
```

**Comment by adamdehaven** [comment]
@IOriens @ulivz What is shown here is not a valid solution. 

The canonical link gets added to the `<head>` of the document for the initial page load; however, the tag never gets updated on subsequent navigation throughout the VuePress SPA. Whatever the `href` property is set to on the first page loaded remains the value as the user navigates to other pages.

**Comment by pontius-g** [comment]
initialy, canonical meta-tag was needed for search-engine indexing purpose which is irrelevant for SPA behavior.
that is why it was performed on server-side rendering and has no issue for further real-client page navigation

**Comment by adamdehaven** [comment]
@pontius-g No argument as to the purpose of the canonical tag - but link tags are used as a web standard not only for SEO, but for rendering other JavaScript content (e.g. AMP pages). Hoping a future VuePress update will allow for injecting these tags as needed. I [posted a start to a possible change that could be implemented](https://github.com/vuejs/vuepress/issues/2153#issuecomment-669602423) on another issue.

**Comment by 2color** [comment]
I'm not an expert on Vuepress or Vue, but I stumbled upon this while trying to add default canonical tag.

I ended up writing a small custom plugin that automatically adds the canonical tag if it's not defined in the frontmatter. It relies on the https://vuepress.vuejs.org/guide/frontmatter.html#canonicalurl config which is available from 1.7.1

```js
module.exports = (options, context) => ({
  name: 'vuepress-default-canonical',
  extendPageData($page) {
    const { frontmatter } = $page
    // If no canonicalUrl is explictly defined in the Frontmatter, add it based on the permaLink
    if (!frontmatter.canonicalUrl && frontmatter.permalink) {
      frontmatter.canonicalUrl = `${$page?._context?.themeConfig?.domain}${frontmatter.permalink}`
    }

    if (!frontmatter.permalink) {
      if ($page._permalink === '/') {
        // Set the canonical URL to the root page (which doesn't have a frontmatter permalink)
        frontmatter.canonicalUrl = `${$page?._context?.themeConfig?.domain}`
      }
      return
    }
  },
})
```
